Setup logical replication

sravfeyn opened the pull request:
## Product Description

Utils for setting up logical replication

https://docs.google.com/document/d/1r9uFUQK8cOreaUm6E8nx9nyeKZL8tcNLJwJiWXh8kGI/edit?tab=t.0#heading=h.igbj18ssbv2z

## Technical Summary

- Adds a `SECONDARY_DATABASE_URL` option to local env
- Adds `migrate_multi` and `setup_logical_replication` commands to setup replication

## Safety Assurance

### Safety story

I have tested this locally and will test on staging

### Automated test coverage
NA

### Labels & Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

sravfeyn commented on docker/start_migrate:
```diff
@@ -6,6 +6,6 @@ set -o nounset
 
 echo "Django migrate"
-python manage.py migrate --noinput
+python manage.py migrate_multi --noinput
```
It's simpler to have `migrate_multi` than conditionally run `migrate  --database=secondary`

sravfeyn commented on config/db_router.py:
```diff
@@ -0,0 +1,63 @@
+        if db == DEFAULT_DB_ALIAS:
+            return True
+        elif db == settings.SECONDARY_DB_ALIAS:
+            # Data migrations using RunPython don't need to be
```
Note we do need all the table schemas (not just the superset tables) even though we enable replication for just  [REPLICATION_ALLOWED_MODELS](https://github.com/dimagi/commcare-connect/pull/481/files#diff-f040749dbea6f228246cc9aead49904004997f8f6212981561e1d4d27ccd91a2R18). This is because django migrations can't work without all schemas in place.

Another notable thing here is that data migrations targetted using `RunPython` module won't be propagated to secondary database as data is replicated at database level.

sravfeyn commented:
@calellowitz I have tested this locally and added few comments wherever this diverged (only slightly) from the spec.

sravfeyn commented on commcare_connect/opportunity/management/commands/setup_logical_replication.py:
```diff
@@ -0,0 +1,120 @@
+]
+
+
+class Command(BaseCommand):
```
I have set this up as a management command instead of a migration as that allows us to run this only on envs where it's needed.

sravfeyn commented:
@calellowitz Here are few scenarios I tested locally so far, let me know if you can think of any more

1. Test basic replication;  initial-sync/update/delete from primary propagates to secondary DB
2. Run a migration with data migration in it, make sure only the schema part of the migration is applied on secondary DB and data update occurs from subsequent replication from main

sravfeyn commented:
Bumping this one, I will be able to move forward after an initial review from you @calellowitz

calellowitz commented on config/settings/base.py:
```diff
@@ -33,6 +33,14 @@
 DATABASES["default"]["ATOMIC_REQUESTS"] = True
 DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
 
+SECONDARY_DB_ALIAS = "default"
```
I don't think we should use a default value that represents a broken setup. It should either be `None` or a valid alias.

calellowitz replied:
Should this be in the env?

sravfeyn replied:
None makes sense. It doesn't need to be in the env, since it can be inferred based on `SECONDARY_DATABASE_URL`

calellowitz commented on config/settings/base.py:
```diff
@@ -33,6 +33,14 @@
 
+SECONDARY_DB_ALIAS = "default"
+
+if env.db("SECONDARY_DATABASE_URL", default=""):
```
should this look up `env` rather than `env.db`?

sravfeyn replied:
env.db is just a util to read the flat URL style spec for DB, right?

calellowitz replied:
Yeah, it transforms the url into a database dict, but there is no reason to do that operation since we are throwing away the result? It isn't particularly expensive but doesn't buy us anything and throws warnings when it hits the default value. We just care if the key exists, which `env` does without `.db`

sravfeyn replied:
It's easier to define it in this way, with less num of vars to keep track of (like PASSWORD, DB, USERNAME etc). And the current primary DB is also setup that way (so is inconsistent). Do you have strong feelings against this?

calellowitz replied:
I'm not sure I understand, its exactly the same number of vars, I'm just asking for this line to remove `.db`. Definitely use `.db` when we setup the actual dict lower down, just not for the existence check (since that isn't what it is meant for, and is what the plain `env` call does).

calellowitz commented on config/settings/base.py:
```diff
@@ -33,6 +33,14 @@
+SECONDARY_DB_ALIAS = "default"
+
+if env.db("SECONDARY_DATABASE_URL", default=""):
+    SECONDARY_DB_ALIAS = "secondary"
```
this should probably come from the env and be set in line 36. then it can be used in the conditional in line 38

sravfeyn replied:
Raised and responded in the [above comment](https://github.com/dimagi/commcare-connect/pull/481/files#r1934098135)

calellowitz commented on config/settings/base.py:
```diff
@@ -33,6 +33,14 @@
+
+if env.db("SECONDARY_DATABASE_URL", default=""):
+    SECONDARY_DB_ALIAS = "secondary"
+    DATABASES[SECONDARY_DB_ALIAS] = env.db("SECONDARY_DATABASE_URL")
```
Nit: this would be easier to find if it were next to the other database line so all the databases defined are right next to each other. You can move the conditional settings above the database dict and do it all in one place.

calellowitz commented on config/db_router.py:
```diff
@@ -0,0 +1,63 @@
+from django.db import DEFAULT_DB_ALIAS
+
+
+class SecondaryDatabaseRouter:
```
This name and docstring are pretty confusing because they imply that this is the router for the secondary db, but it is in fact the router for _all_ DBs.

sravfeyn replied:
I will change it. Though, it is possible to have multiple routers as well.

sravfeyn replied:
Updated

calellowitz reviewed (commented):
Left a few comments and suggestions as well as a few questions. It would be great to have tests for this as well

sravfeyn commented on commcare_connect/opportunity/management/commands/migrate_multi.py:
```diff
@@ -0,0 +1,85 @@
+import sys
```
Yeah, I tried putting these originally in the `utils` dir, but that is not an app so it didn't recognize these commands. Any suggestions for other place? Or shall I just create a new app for this?

calellowitz replied:
I would say a new app, since this is not related to any of our existing apps, which I thik is better than utils anyway.

sravfeyn replied:
This is addressed

sravfeyn commented on commcare_connect/opportunity/management/commands/migrate_multi.py:
```diff
@@ -0,0 +1,85 @@
+        if migration_name is not None:
+            args.append(migration_name)
+
+        options["verbosity"] = 0
```
This overrides the default `verbosity` of 1 to 0 of the command.

calellowitz replied:
why is that preferable? and what impact does that have on what is output? Just curious since our current migrate uses the default verbosity, I was wondering why the change here

sravfeyn replied:
To be honest, this is the default on commcare-hq, so I trusted it makes a sensible default, we can tweak it latter, if it doesn't work well for us.

sravfeyn replied:
I removed this.

sravfeyn commented on commcare_connect/opportunity/management/commands/setup_logical_replication.py:
```diff
@@ -0,0 +1,120 @@
+)
+from commcare_connect.users.models import User
+
+REPLICATION_ALLOWED_MODELS = [
```
Added the two models, and alphabetized them.

Re. making it a setting, what's the case for why these should be 'settable'?

calellowitz replied:
I imagine the list will grow over time, and settings seemed like a logical place to have a list that represents the current state of our configuration, rather than a management command. I could also see it somewhere like the new app you are making, since it won't be configured by environment, either seems fine, but a management command as the source of truth feels bit of a weirder place to look.

calellowitz replied:
Still interested in this question

sravfeyn replied:
I see, you mean to declare it as a constant. I will add a constants file and move it into there.

sravfeyn commented on commcare_connect/opportunity/management/commands/setup_logical_replication.py:
```diff
@@ -0,0 +1,120 @@
+    LearnModule,
+    CompletedModule,
+    Payment,
+    User,
```
It looks like there is no native way (using replication) to do this, I can think of few ways that I am testing
1. Add a delete password column trigger on secondary DB (this should immediately delete the password on primary)
2. Remove the column read permission for users on secondary DB

calellowitz replied:
ok, thanks for looking. that should be fine for now. we can look more into it later

sravfeyn commented on commcare_connect/opportunity/management/commands/migrate_multi.py:
```diff
@@ -0,0 +1,85 @@
+            call_options["database"] = db_alias
+            call_command("migrate", *args, **call_options)
+
+        dbs_to_migrate = [
+            db_alias for db_alias in settings.DATABASES.keys() if settings.DATABASES[db_alias].get("MIGRATE", True)
+        ]
+        dbs_to_skip = list(set(settings.DATABASES) - set(dbs_to_migrate))
```
Removed

sravfeyn commented on commcare_connect/opportunity/management/commands/migrate_multi.py:
```diff
@@ -0,0 +1,85 @@
+        if dbs_to_skip:
+            print("\nThe following databases will be skipped:\n * {}\n".format("\n * ".join(dbs_to_skip)))
+
+        jobs = [gevent.spawn(migrate_db, db_alias) for db_alias in dbs_to_migrate]
```
Good point, I will test this out

sravfeyn commented on commcare_connect/opportunity/management/commands/setup_logical_replication.py:
```diff
@@ -0,0 +1,120 @@
+            try:
+                # Check if publication exists
+                cursor.execute("SELECT pubname FROM pg_publication WHERE pubname = %s;", [publication_name])
+                if cursor.fetchone():
```
It involves couple of commands like (`ALTER PUBLICATION`) that can be written as command like this one.

calellowitz replied:
it would be great if this command could be updated to handle that, so that we don't have to write new custom commands each time. That would likely also make it resilient to failure part way through, if we need to restart it. Basically an idempotent command that takes a list of all tables we want in that state seems ideal.

calellowitz replied:
Still interested in this question

sravfeyn replied:
(this was part of hidden comments, so I din't notice)

I didn't mean a command each time, it would be sort of easy to make this command idempotent. There is something called `pg_replication_tables` function which gives the list of currently publicated tables, so it's just a matter of getting the diff and running the SQL command to add/remove the diff tables.

I didn't make it's easy to do it when actually needed instead of the extra effort now.

Also, to make sure I capture all the tables from current superset, I ran a script to pull all queries from superset (using their API) and parse the list of tables. So, I am a bit more confident that we won't need that command in the short term.

sravfeyn commented on commcare_connect/opportunity/management/commands/setup_logical_replication.py:
```diff
@@ -0,0 +1,120 @@
+                        """
+                    )
+                    self.stdout.write(self.style.SUCCESS(f"Subscription '{subscription_name}' created successfully."))
+            except Exception as e:
```
Good point, removed.

sravfeyn commented on config/db_router.py:
```diff
@@ -0,0 +1,63 @@
+            #   applied on secondary DB as they are replicated
+            #   at the database level.
+            operation = hints.get("operation", None)
+            if operation is None:
```
I haven't examined closely, but I came across various checks that called this function with operation None.

sravfeyn commented on config/db_router.py:
```diff
@@ -0,0 +1,63 @@
+            if operation is None:
+                return True
+
+            # Allow schema-only operations
```
These are all the possible schema changing operations, though I do have to examine these again to be sure.

There is `RunSQL` but that we don't use it so far and use-case for it is usually not schema migrations.

sravfeyn replied:
I already tried blacklisting RunPython only, but that didn't work well. 

I will re-look at better ways of managing this, since it also has to be future proof. (for e.g. when a dev does have to add a RunSQL to update some schema change that can't be done using ORM)

calellowitz replied:
If we included every operation in allowed operations and always return True, why not just make it `return True`

sravfeyn replied:
This comment addresses this https://github.com/dimagi/commcare-connect/pull/481#discussion_r1948945459

sravfeyn commented:
@calellowitz Thanks for the comments. At what point do we want to test this on staging?

calellowitz commented:
I think I mixed standalone comments and a review by mistake there, but tried to get to your questions. This looks fine to test on staging now, since my remaining comments won't effect behavior or are about what happens if we make later changes to the db list, which we will need to test after rollout anyway.

However, I don't see how you plan to setup the second database anywhere, what kinds of permissions you plan to give the user, and how. Ideally we would want to review those as well before rolling this out.

sravfeyn commented:
Here is what I am thinking re.staging
- Create a new RDS instance in our AWS CCC account
- Setup logical replication using this PR (in the public schema)
- Create a readonly user that has `readonly` role on public schema and a `write` permissions on a separate schema called `extended`
- Use this user inside superset.

sravfeyn commented:
@calellowitz 

1. I have addressed your comments
2. Tested this on staging (sent you an output of the replication status command). 
3. Added rollout instructions to the [google doc](https://docs.google.com/document/d/1r9uFUQK8cOreaUm6E8nx9nyeKZL8tcNLJwJiWXh8kGI/edit?tab=t.0#heading=h.9u8erzkceulj) under Production rollout

I have tested the replication part, but didn't test the superset database switching part since we don't have staging-superset. But I think this is straighforward and short process

calellowitz commented on commcare_connect/multidb/management/commands/setup_logical_replication.py:
```diff
@@ -68,53 +79,48 @@ def handle(self, *args, **options):
+                    f"dbname={default_db_settings['NAME']} "
+                    f"user={default_db_settings['USER']} "
+                    f"password={default_db_settings['PASSWORD']} "
+                    f"sslmode=disable"  # same VPC in AWS
```
why disable SSL, especially if we don't for any of the other actions?

sravfeyn replied:
This is undone, it was one of the things I tried.

calellowitz commented on commcare_connect/multidb/db_router.py:
```diff
@@ -0,0 +1,31 @@
+    def allow_migrate(self, db, app_label, model_name=None, **hints):
+        if db == DEFAULT_DB_ALIAS:
+            return True
+        elif db and db == settings.SECONDARY_DB_ALIAS:
```
Rather than requiring this for `RunPython`, can we just default it to false in that case? That way no one has to remember to add it and the normal behavior is used. The validation you added is a nice backup option but it would be great if we didn't have to add this everywhere.

sravfeyn replied:
This has been a tricky thing to get it working. 

The allow_migrate actually has no visibility into what kind of operation (whether RunPython/CreateModel/AddField etc) unless the hints are passed from the operation itself. So we would needed to have an hint called `operation=RunPython`. But instead of passing operation, I have added `run_on_secondary` which can be passed from any migration, though for RunPython, I made it mandatory.

calellowitz reviewed (commented):
one new small question generally looks, besides the old questions I pinged on again

calellowitz commented on commcare_connect/opportunity/management/commands/setup_logical_replication.py:
```diff
@@ -0,0 +1,120 @@
+                    )
+                else:
+                    # Create new subscription
+                    default_db_settings = default_conn.settings_dict
```
Still interested in this question

sravfeyn replied:
Yeah, I was planning to do the same. The user on primary should be granted only the `REPLICATIOn` privilege.  I have modified the code to get those creds via standin with a message.

```
self.stdout.write("Provide user credentials on primary with only replication privilege")
username = input("Enter username: ")
password = input("Enter password: ")
```

sravfeyn commented:
@calellowitz Bumping this if this is good to go. I have also responded to your questions on the google doc.
